gpii/node_modules/flowManager/src/RevisionGetHandler.js

```diff
@@ -0,0 +1,55 @@
+/*!
+GPII GET /revision handler used by the Cloud Based Flowmanager
+
+Copyright 2020 OCAD University
+
+Licensed under the New BSD license. You may not use this file except in
+compliance with this License.
+
+You may obtain a copy of the License at
+https://github.com/GPII/universal/blob/master/LICENSE.txt
+*/
+
+"use strict";
+
+var fs = require("fs"),
+    fluid = require("infusion"),
+    gpii = fluid.registerNamespace("gpii");
+
+fluid.defaults("gpii.flowManager.cloudBased.gpiiRevision.handler", {
+    gradeNames: ["kettle.request.http"],
+    invokers: {
+        handleRequest: {
+            funcName: "gpii.flowManager.cloudBased.gpiiRevision.handler.handleRequest",
+            args: ["{gpii.flowManager.cloudBased}.options.gpiiRevisionPath", "{that}"]
+        }
+    }
+});
+
+gpii.flowManager.cloudBased.gpiiRevision.handler.reportError = function (that, filePath, errMsg) {
+    fluid.log ("Error retrieving full git revision: '" + filePath + "'");
+    that.events.onError.fire({
+        isError: true,
+        message: "Error retrieving full git revision: " + errMsg,
+        statusCode: 404
+    });
+};
+
+gpii.flowManager.cloudBased.gpiiRevision.handler.handleRequest = function (gpiiRevisionPath, that) {
+    var errMsg;
+    var fullPath = fluid.module.resolvePath(gpiiRevisionPath);
+    var result = {sha256: ""};
+    try {
+        result = JSON.parse(fs.readFileSync(fullPath, "utf-8"));
```

Comment:
It would be worthwhile to check the existence of `fullPath` before reading it and expecting the `catch` to report the error.

As `gpii-revision.json` doesn't exist in the universal root directory, right now running tests populates the log with fatal errors although tests will complete successfully:
```
15:30:13.955:  FATAL ERROR: Uncaught exception: Error retrieving full git revision: ENOENT: no such file or directory, open '/home/vagrant/sync/universal/gpii-revision.json'
undefined
...
```

Reply:
@cindyli This is actually part of PR #837 that has only the changes for the CBFM `/revision` endpoint.  I'll fix it there and merge it into this one.

Reply:
I started on this, but there is a bigger issue. 

This is the `/revision` GET handler, and like other GET handlers, it is either successful, or there is an error, here a `404`.  The success/failure is implemented in the handler by firing either an `onSuccess` or `onError` event.

Thus, checking if the file exists doesn't help, since a `200` (success) is not correct if the file is missing. The `onError` event is still going to fire, and will result in a "`FATAL ERROR: ...`" message in the tests.

I'm wondering if this is peculiar to the tests, and whetehr there is a way to avoid the error for cases where getting the revision is not relevant to a specific test.

Reply:
It seems [`fs.access`](https://nodejs.org/api/fs.html#fs_fs_access_path_mode_callback) or `fs.accessSync` check the file existence without opening it, so they don't report fatal errors. Please give a try. Thanks.

Reply:
Using `fs.accessSync()` doesn't help.

The _reporting_ of the `FATAL ERROR: ...` is not caused by the failure of `fs.readFileSync()` or either of the access functions.  It's because that failure is relayed to an `onError` event a few lines later, using the call to `gpii.flowManager.cloudBased.gpiiRevision.handler.reportError()`.

Both of the "missing revision file" and "malformed revision file" errors result in a `FATAL...` log message.  The latter produces messages like:
> `17:34:19.500:  FATAL ERROR: Uncaught exception: Error retrieving full git revision: Missing revision value`

Note that both error conditions are tested in the `CloudRevisionGetTests.js` in the test cases [`testInvalidRevision`](https://github.com/klown/universal/blob/GPII-4273iii/tests/platform/cloud/CloudRevisionGetTests.js#L86) and [`testMissingRevision`](https://github.com/klown/universal/blob/GPII-4273iii/tests/platform/cloud/CloudRevisionGetTests.js#L98)

The use of `onError` for requests is used in many places in the GPII.  For example, the `/ready` and `/health` endpoints output `FATAL...` log messages when they fail, and there are others.  The reason there are so many in the case of `/revision` is because the new local SRDS always requests the revision when it initializes, and gpii-universal doesn't have such a file in the repository. So, for every test that causes the creation of a local SRDS, this error will occur. Which is, admittedly, a lot.  But, as you noted, it's harmless in that is does not interfere with any given test, just like all the other instances of `FATAL ...` log messages.

I'm exploring if there is a way to configure the tests such that revision requests are omitted where that's not important to the tests.

Reply:
Good point of omitting revision requests in tests. This probably makes sense for running the system in dev configs too, in which case, people or developers who are testing the system might prefer to read the in-dev solutions registry directly from the file system.

My thought is:
1. Move [configs for the SR remote fetch](https://github.com/klown/universal/blob/GPII-4273iii/gpii/node_modules/flowManager/configs/gpii.flowManager.config.local.base.json5#L9-L20) out of the `gpii.flowManager.config.local.base` into its own separate config file. This config will included into production configs in "gpii/configs/" directory. An example of such a separate config is [here](https://github.com/klown/universal/blob/GPII-4273iii/gpii/node_modules/flowManager/configs/gpii.flowManager.config.suppressHttpEndpoints.base.json5). If we do so, a documentation should be added for this config at [configs.md](https://github.com/klown/universal/blob/GPII-4273iii/documentation/Configs.md#special-configs);

2. In this separate config, the distribution target for `cloudURL`, `urlPrefix`, `urlSuffix` will be `gpii.flowManager.solutionsRegistry.dataSource.local` rather than its sub-components `revisionRequester` and `repositorySolutionsLoader`, so that [`loadSolutions.getRevision` listener](https://github.com/klown/universal/blob/GPII-4273iii/gpii/node_modules/flowManager/src/SolutionsRegistryDataSource.js#L174-L177) can verify `cloudURL !== null` before sending requests to CBFM and [`loadSolutions.loadFromRepository`] can verify `urlPrefix` before fetching from Github repo. Extra `distributeOptions` blocks will be added to `gpii.flowManager.solutionsRegistry.dataSource.local` to distribute these 3 values down to its subcomponents;

3. Add a test config for testing SR fetch from Github repo.

Let me know if you come up with better solutions. Thanks.

Reply:
The "problem" with the `FATAL ...` messages in the log for some of the tests
is twofold:

1. Some of the tests instatiate a CBFM that supports the `/revision` endpoint.  AND they instantiate a LFM that uses the new `SRDS` that, in turn, queries that CBFM `/revision` endpoint.  The CBFM's revision request handler will fire an `onError` event in this case because there is no revision file for the CBFM to return.  The error event is surfaced as a `FATAL ...` error message.  This is the way it is supposed to work.  For example, by comparison, the production tests do not show the error because they are set up to mimic a production system and, as such, have a revision file.  If the revision file was missing or malformed in a production scenario, you'd want to see the `FATAL` error messages.

2. The testing framework defines a fluid component whose job is to ensure that `FATAL` errors are logged.  The comment about the [`errorLoggingRequest` component](https://github.com/GPII/universal/blob/master/gpii/node_modules/testing/index.js#L8) indicates that this is a requirement: 

> `Some definitions to make sure that errors during request get logged to the console. This should really be a default`

I removed that component as an experiment and none of the `FATAL` error messages are logged, but otherwise everything else behaves as before.

The dilemma is that the testing framework sees the `FATAL` error messages as required, but the ones caused by a CBFM-with-revision-endpoint + a LFM-with-new-SRDS are irrelevant for the majority of the tests.

The compromise I came up with is to defeat the revision request, but in just those cases where it is pointless.  This is done by changing a couple of the configuration file to misdirect the `/revision` request away from the CBFM.  That is, when the SRDS asks for the revision, it asks that of a server that has no `/revision` endpoint and, in pariticular, no revision request handler.  The `onError` event won't occur.  However, for the specific tests of the `/revision` endpoint (`CloudRevisionGetTests.js`, and the production tests), the configuration is correctly set and the `FATAL` messages are output.  Furthermore, any other tests that do incur a `FATAL` error continue to work as before.  Those `FATAL` errors are logged.